modules: deprecate module.parent

aduh95 opened the pull request:
This feature does not work when a module is imported using ECMAScript modules
specification, therefore it is deprecated.

This has been discussed during [Node.js Modules Team Meeting 2020-03-11](https://docs.google.com/document/d/1KZBgLnJn3ZejOIitOqTylXwKdCBZ7Byka9T3fUch3wU/view); the main issue with this feature is that users are using it to check if the current module is the entry point, but that breaks when a CJS module is imported using ESM loader. Also its behaviour is kind of counterintuitive because of CJS cache.

This PR contains a breaking change: `Object.keys(module).includes('parent')` used to be `true`, and is now `false`. Hopefully this can get merged before v14 feature freeze.

Fixes: https://github.com/nodejs/modules/issues/469

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

mscdex commented on doc/api/deprecations.md:
````diff
@@ -2617,6 +2617,30 @@ async function openAndClose() {
+Instead, it is recommended to compare `require.main` and `module`:
+
+```js
+if (require.main === module) {
````
Would `process.mainModule === module` be a better choice?

bmeck replied:
yes per https://github.com/nodejs/node/blob/6de7b635a6185115975248efbc71fb8c205b59b6/doc/api/modules.md#accessing-the-main-module and https://github.com/nodejs/node/blob/6de7b635a6185115975248efbc71fb8c205b59b6/doc/api/process.md#processmainmodule usage of process.mainModule and require.main is likely preferred (though note, those values may be undefined)

aduh95 replied:
`process.mainModule === undefined` when run from an ES module, while `require.main` throws. I'd say it's probably a better idea to promote a pattern that is clearly meant for CJS. If you get `import.meta.main` for ESM, I would expect `process.mainModule` to be deprecated.

Just my opinion though, if you feel otherwise I can update the PR.

DerekNonGeneric commented:
### ❌I'm opposed

My dissent is voiced in a comment to the related issue.

https://github.com/nodejs/modules/issues/469#issuecomment-598067025

bmeck commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -43,6 +43,7 @@ const {
   StringPrototypeMatch,
   StringPrototypeSlice,
   StringPrototypeStartsWith,
+  WeakMap,
```
please use `SafeWeakMap` off primordials

bmeck commented:
While we are here, can we update https://github.com/nodejs/node/blob/master/doc/api/modules.md#moduleparent to state that the value may be `undefined` or `null`. Regardless of the deprecation logic, that is a driving force behind this PR.

bmeck commented:
@addaleax do you feel comfortable with docs deprecation?

aduh95 commented:
@addaleax it breaks when people are using ESM or REPL, there is a better alternative that always works, and it seems there is no actual use-case for this feature.

Anyway, would you find it acceptable if the warning is displayed only when `module.parent` is `undefined` or `null`? This way the warning is displayed when the feature may break, hence helping the debug process.

jasnell commented on doc/api/deprecations.md:
````diff
@@ -2617,6 +2617,49 @@ async function openAndClose() {
 ```
 
+<a id="DEP0138"></a>
+### DEP0138: `module.parent`
````
```suggestion
<a id="DEP0XXX"></a>
### DEP0XXX: `module.parent`
```

Number assigned when landing

jasnell commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -226,6 +228,17 @@ ObjectDefineProperty(Module, 'wrapper', {
+    },
+    'Module.prototype.parent is deprecated and may be removed in a later ' +
+    'version of Node.js.',
+    'DEP0138'
```
```suggestion
    'DEP0XXX'
```

jasnell commented on test/parallel/test-module-parent-deprecation.js:
```diff
@@ -0,0 +1,12 @@
+  'DeprecationWarning',
+  'Module.prototype.parent is deprecated and may be removed in a later ' +
+  'version of Node.js.',
+  'DEP0138'
```
```suggestion
  'DEP0XXX'
```

aduh95 commented:
> When it does break, it will most likely break in a way that’s easily noticed.

@addaleax I don't think that's a fair assumption: when I did encounter the issue, I was porting a library of mine to ESM, and it was not obvious what what did go wrong. One of the dependency is using a check on `module.parent` to run its tests, tests that have side effects but there were no console output indicating tests were being run. Even reading the dependency source code, it wasn't obvious what was causing the issue, I didn't know `module.parent` would be `undefined` because I was importing from ESM.

I know this is just anecdotal evidence, yet I am convinced a runtime warning is needed; hopefully my user story shows how Node.js can benefit from a runtime deprecation.

To avoid leaking warnings to library users, I have changed the algorithm to warn only when `module.parent` is `null` or `undefined`. My reasoning behind this choice is:

- Users that have a dependency checking `module.parent` won't get the warning when using `require` (because it works fine in that situation)
- Users that have a dependency checking `module.parent` will get the warning when using `import` (because it breaks in that situation)
- Library authors checking `module.parent` to run their tests will get the warning so they know they should fix their code.
- Users using a CLI package that checks `module.parent` will also get the warning, that's unfortunate but I would expect it is not a very common pattern.

I have also added a code snippet in the deprecation note to tell users how to get the current module's parents to address @DerekNonGeneric's concern.

devsnek commented on doc/api/deprecations.md:
````diff
@@ -2617,6 +2617,49 @@ async function openAndClose() {
+
+```js
+const moduleParents = [];
+function findCurrentModuleParents(moduleParent = require.main) {
````
this code will form an infinite loop in the case of circular dependencies. I would recommend doing something like this:
```js
const parents = Object.values(require.cache)
  .filter((m) => m.children.includes(module));
```

aduh95 replied:
That's way more concise and readable, thanks!

addaleax commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -226,6 +228,23 @@ ObjectDefineProperty(Module, 'wrapper', {
+      warned = true;
+      process.emitWarning(
+        'module.parent is deprecated and may be removed in a later ' +
+          'version of Node.js.',
```
I would be more explicit about what this warns against – something stating that `module.parent` may not always be accurate. Removing it is something that I don’t hope anybody’s suggesting at this point?

aduh95 replied:
I would bet that this will never be removed (probably ever), I just copied another deprecation warning and changed the function name. Would you be OK with the following phrasing?

```suggestion
        'module.parent is deprecated and is know to break when using ' +
          'ECMAScript module specification. Consider replacing it with ' +
          'require.main === module.',
```

jasnell commented on doc/api/deprecations.md:
```diff
@@ -2617,6 +2617,38 @@ async function openAndClose() {
+
+Type: Runtime
+
+A CJS module can access the first module that required it using `module.parent`.
```
Nit: recommend s/CJS/CommonJS

ExE-Boss commented on doc/api/deprecations.md:
```diff
@@ -2623,14 +2623,14 @@ async function openAndClose() {
-specification, therefore it is deprecated.
+A CommonJS module can access the first module that required it using
+`module.parent`. This feature does not work when a module is imported using
+ECMAScript modules specification, therefore it is deprecated.
```
The grammar structure here feels weird, particularly the use of “imported using ECMAScript modules specification”.

ExE-Boss commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -228,25 +228,16 @@ ObjectDefineProperty(Module, 'wrapper', {
+  get: pendingDeprecation ? deprecate(
+    getModuleParent,
+    'module.parent is deprecated due to accuracy issues. Please use ' +
+            'require.main === module to check for CLI usage instead.',
```
This indentation doesn’t feel right.

ExE-Boss commented on doc/api/deprecations.md:
```diff
@@ -2617,6 +2617,38 @@ async function openAndClose() {
+
+A CommonJS module can access the first module that required it using
+`module.parent`. This feature does not work when a module is imported using
+ECMAScript modules specification, therefore it is deprecated.
```
The grammar structure here feels weird, particularly the use of “imported using ECMAScript modules specification”.

aduh95 replied:
Both static `import` and dynamic `import()` trigger the bug. I am not a native English speaker, I'm definitely open to suggestions to improve my grammar or the phrasing in general :)

devsnek replied:
"this feature is being deprecated because it does not work consistently in the presence of ECMAScript modules and because it gives an inaccurate representation of the CommonJS module graph"

aduh95 replied:
```suggestion
A CommonJS module can access the first module that required it using
`module.parent`. This feature is deprecated because it does not work
consistently in the presence of ECMAScript modules and because it gives an
inaccurate representation of the CommonJS module graph.
```

ExE-Boss commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -226,6 +228,18 @@ ObjectDefineProperty(Module, 'wrapper', {
+  get: pendingDeprecation ? deprecate(
+    getModuleParent,
+    'module.parent is deprecated due to accuracy issues. Please use ' +
+            'require.main === module to check for CLI usage instead.',
```
This indentation doesn’t feel right.

aduh95 replied:
Good catch, apparently `make lint-js-fix` doesn't fix everything...

```suggestion
    'module.parent is deprecated due to accuracy issues. Please use ' +
      'require.main === module to check for CLI usage instead.',
```

jasnell commented on doc/api/deprecations.md:
````diff
@@ -2617,6 +2617,39 @@ async function openAndClose() {
+
+```js
+if (require.main === module) {
+  // CLI code runs here
````
Given that nothing in the text mentions CLI's, this comment is a bit confusing.

aduh95 replied:
```suggestion
  // Code section that will run only if current file is the entry point.
```

jasnell commented on test/parallel/test-module-parent-deprecation.js:
```diff
@@ -0,0 +1,14 @@
+common.expectWarning(
+  'DeprecationWarning',
+  'module.parent is deprecated due to accuracy issues. Please use ' +
+          'require.main === module to check for CLI usage instead.',
```
The mention of CLI here is a bit confusing.

aduh95 replied:
```suggestion
    'require.main to find program entry point instead.',
```

aduh95 commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -226,6 +228,18 @@ ObjectDefineProperty(Module, 'wrapper', {
+  get: pendingDeprecation ? deprecate(
+    getModuleParent,
+    'module.parent is deprecated due to accuracy issues. Please use ' +
+      'require.main === module to check for CLI usage instead.',
```
```suggestion
      'require.main to find program entry point instead.',
```

aduh95 commented:
@bmeck could have a look to check if you're content with my changes?

aduh95 commented:
This PR would need another approval from @nodejs/tsc in order to land because it is a breaking change. Maybe @jasnell, since you've reviewed it already?

addaleax commented:
@nodejs/tsc ^^^

addaleax commented:
This needs a rebase, it seems.

dgellow commented on doc/api/deprecations.md:
````diff
@@ -2651,6 +2651,39 @@ written twice. This introduces a race condition between threads, and is a
+}
+```
+
+When looking for the CJS modules that have required the current one,
````
If I may ask: how do you do that from an ECMAScript module?

aduh95 replied:
I don't think their is an API available from a module itself; however if you are looking for a way to show the dependency graph, I think it can be achieved using [loader hooks](https://nodejs.org/docs/latest-v13.x/api/esm.html#esm_hooks).

dgellow replied:
Surely not the best approach, but I found that by creating an error and inspecting the stack trace I could find out what modules imported and called the current one.

jasnell commented:
Still LGTM

aduh95 commented:
@nodejs-github-bot test this

aduh95 commented:
I have re-rebased to fix the merge conflict. Is there something else I can do to have it ready to land?

addaleax commented on doc/api/deprecations.md:
```diff
@@ -2644,7 +2644,7 @@ changes:
 -->
 
-Type: Runtime
+Type: Documentation-only (supports [`--pending-deprecation`][])
```
Is this changing the text for the wrong entry?

aduh95 replied:
Good catch! Thanks @addaleax!

tniessen commented:
@DerekNonGeneric I believe suitable alternatives exist. Is there something `module.parent` can do that other features cannot achieve in a similar/better way?

ExE-Boss commented:
@tniessen
> Is there something `module.parent` can do that other features cannot achieve in a similar/better way?

Basically, no:

> When looking for the CommonJS modules that have required the current one,
`require.cache` and `module.children` can be used:
>
> ```js
> const moduleParents = Object.values(require.cache)
> 	.filter(m => m.children.includes(module));
> ```

---

The only thing `module.parent` leaks is whether the module was first imported by an **ECMAScript** or **CommonJS** module.

DerekNonGeneric commented:
/to @tniessen

I have nothing more to contribute to this discussion beyond my remarks in the associated issue. We can always revisit this if someone proposes it become a runtime deprecation, but until then, and out of respect to those in favor, I'll be seeing myself out of this PR thread.

![homer-bushes](https://user-images.githubusercontent.com/17770407/81519974-eb05b100-9310-11ea-99d1-6ec977e28f92.gif)

aduh95 commented:
CI is green,
Violets are blue,
I would very much like
See this PR go through! 😅

BridgeAR commented:
I updated the semverness, since documentation deprecations do not count as semver-major (as outlines in our guidelines).

BridgeAR commented:
Landed in 6443ab9595de

aduh95 commented:
Thank you for landing this! Should this get a `Don't land on v14.x` label? It does mutate the `module` object in order to enable the `--pending-deprecation` warning, I'm not sure if that may be breaking or not… I don't mind working on a backport without the `--pending-deprecation` trick if that helps.

BridgeAR commented:
@aduh95 I marked it with the backport request label. That way we do not pull this PR in automatically. If you open a backport without the change, that would be great.

MylesBorins commented on doc/api/deprecations.md:
```diff
@@ -2697,6 +2697,39 @@ The `repl` module exports a `_builtinLibs` property that contains an array with
 native modules. It was incomplete so far and instead it's better to rely upon
 `require('module').builtinModules`.
 
+<a id="DEP0143"></a>
```
Realizing in retrospect taht this should have landed without an assigned deprecation number, this should happen at the time of release.

aduh95 replied:
FWIW this PR was using a placeholder originally, and I changed it to the deprecation number based on https://github.com/nodejs/node/pull/33430#issuecomment-629626771

richardlau replied:
Deprecation codes are supposed to be assigned when a PR lands: https://github.com/nodejs/node/blob/master/doc/guides/releases.md#step-3-update-any-replaceme-and-dep00xx-tags-in-the-docs

papb commented:
A bit late to the party, but [meow](https://github.com/sindresorhus/meow/) uses this for a reason other than checking if the module is the entrypoint. See https://github.com/sindresorhus/meow/issues/157

bmeck commented:
@papb yes, we have talked about meow and similar for quite some time ( I see a ref from 2017 in https://github.com/nodejs/node-eps/issues/57#issuecomment-300870976 )

richardlau commented:
After bisecting, this PR is the reason that node-gyp's test suite is failing in CITGM on master and the current v15.0.0-proposal (#35014, cc @BethGriggs).
e.g. https://ci.nodejs.org/job/citgm-smoker/2497/nodes=ubuntu1804-64/testReport/junit/(root)/citgm/node_gyp_v7_1_0/
```
 # Subtest: test/test-configure-python.js
     1..0 # no tests found
 not ok 2 - test/test-configure-python.js # time=409.579ms
   ---
   timeout: 120000
   file: test/test-configure-python.js
   childId: 1
   command: /home/iojs/build/workspace/citgm-smoker/nodes/ubuntu1804-64/smoker/bin/node
   args:
     - '-r'
     - >-
       /home/iojs/tmp/citgm_tmp/a30ded27-455a-421a-8399-2875b04856df/node-gyp/node_modules/esm/esm.js
     - test/test-configure-python.js
   stdio:
     - 0
     - pipe
     - 2
   cwd: /home/iojs/tmp/citgm_tmp/a30ded27-455a-421a-8399-2875b04856df/node-gyp
   exitCode: 1
   ...
...
 /home/iojs/tmp/citgm_tmp/a30ded27-455a-421a-8399-2875b04856df/node-gyp/node_modules/require-inject/index.js:72
   Object.keys(mocks).forEach(function (name) {
                      ^
 TypeError: Cannot set property parent of #<Module> which has only a getter
     at /home/iojs/tmp/citgm_tmp/a30ded27-455a-421a-8399-2875b04856df/node-gyp/node_modules/require-inject/index.js:82:18
     at Array.forEach (<anonymous>)
     at installGlobally (/home/iojs/tmp/citgm_tmp/a30ded27-455a-421a-8399-2875b04856df/node-gyp/node_modules/require-inject/index.js:72:22)
     at requireInject (/home/iojs/tmp/citgm_tmp/a30ded27-455a-421a-8399-2875b04856df/node-gyp/node_modules/require-inject/index.js:41:7)
     at module.exports (/home/iojs/tmp/citgm_tmp/a30ded27-455a-421a-8399-2875b04856df/node-gyp/node_modules/require-inject/index.js:13:10)
     at Object.<anonymous> (/home/iojs/tmp/citgm_tmp/a30ded27-455a-421a-8399-2875b04856df/node-gyp/test/test-configure-python.js:8:19)
     at Module._compile (internal/modules/cjs/loader.js:1072:30)
     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
     at Module.load (internal/modules/cjs/loader.js:937:32)
     at Function.Module._load (internal/modules/cjs/loader.js:778:14)
```

The test suite passes on Node.js 14.

@nodejs/modules-active-members is this expected to break for CJS? This was documented as being a documentation only deprecation.

bmeck commented:
@richardlau this looks like whenever the deprecated form landed it didn't add back a setter, to my knowledge there was no intent to prevent assignment in CJS. This looks like a bug and we need to add back a setter, right now https://github.com/iarna/require-inject/blob/b1cc122fa34741a5d60ad4c26e839fd9bbb7a61b/index.js#L36 is failing due to the lack of setter. changing https://github.com/nodejs/node/blob/8beef5eeb82425b13d447b50beafb04ece7f91b1/lib/internal/modules/cjs/loader.js#L226-L233 to have a setter should fix it

aduh95 commented:
This PR moved the `parent` declaration from the `Module` instance to the prototype to ensure the pending deprecation warning is written only once. As a result `module.parent` is now read-only. It's a breaking change, and has been backported to `v14.x` in semver-minor patch.

`require-inject` is apparently creating a `Module` instance, and tries to [assign it a `parent` property](https://github.com/iarna/require-inject/blob/b1cc122fa34741a5d60ad4c26e839fd9bbb7a61b/index.js#L78-L82)  which throws. A possible fix would be to add a setter to `Module.prototype.parent`.

guybedford commented:
FYI this is about to go out in 12.

targos commented:
@guybedford The version of this change that's going out in 12 is https://github.com/nodejs/node/pull/34592 (doc-only deprecation)

guybedford commented:
Ah great, thanks for verifying!

On Tue, Oct 6, 2020 at 10:46 Michaël Zasso <notifications@github.com> wrote:

> @guybedford <https://github.com/guybedford> The version of this change
> that's going out in 12 is #34592
> <https://github.com/nodejs/node/pull/34592> (doc-only deprecation)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/32217#issuecomment-704441030>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAESFSVVXPE47WUCW7UCWGTSJNJXXANCNFSM4LGAHYGQ>
> .
>

edumaciel10 reviewed (commented):
cool
